### ISSUE TITLE
Flattening of extends with self-references

### ISSUE THREAD

**Issue by modelica-trac-importer** [tbeu/ModelicaSpecification#496]
**Reported by perost on 9 Apr 2014 13:19 UTC**
Consider the following model:
```
model T1
  Real x;
end T1;

model T2
  Real x = 0.0;
  Real y = 0.0;
end T2;

model A
  replaceable T1 t;
end A;

model B
  extends A(redeclare T2 t(x = t.y));
end B;
```
The spec says that when flattening B we should look up the base class A, flatten it and insert it into the class. To be able to flatten A we need to apply the modifier on it. The modifier references t.y though, and to look up that name we need to flatten A. Does this case break the rule that it should be possible to flatten all extends-clauses before inserting the result of their flattening, or is there some way of handling this?

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1458

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 9 Apr 2014 13:33 UTC**
Replying to [ticket:1458 perost]:
> The spec says that when flattening B we should look up the base class A, flatten it and insert it into the class. To be able to flatten A we need to apply the modifier on it. 

Technically it says that we need to merge the modifiers to A - not "apply it", i.e. it does not state that we need to evaluate the modifier yet; and thus we can delay it.

In this case the merging is kind of obvious - since the declaration does not have any.

> The modifier references t.y though, and to look up that name we need to flatten A. Does this case break the rule that it should be possible to flatten all extends-clauses before inserting the result of their flattening, or is there some way of handling this?

It shouldn't break that rule - the rule is intended to guard against lookup being messed up, e.g. by having model A with an A inside it.

**Comment by modelica-trac-importer** [comment]
**Comment by adrpo on 11 Apr 2014 05:03 UTC**
> Technically it says that we need to merge the modifiers to A - not "apply it"

Well, the specification text is:

5.6.1 Partial Flattening
First the names of declared local classes and components are found. Here modifiers are merged to the local elements and redeclarations take effect (including redeclared elements, see Section 7.3). 
Then base-classes are looked up, flattened and inserted into the class ...

5.6.2 Flattening
Partially flatten the class, apply the modifiers (Section 7.2) and flatten all local elements.

So 5.6.1 say that base-classes are looked up, flattened and 5.6.2 
say that one has to partially flatten the class, apply the modifiers.

This is a bit wrong as partially flatten the class contains merging of modifiers.
What does "apply the modifers" in 5.6.2 means then?

Basically what you're saying is that modifiers are merged blindly without even checking that they exist (no lookup) until the flattening is done.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 11 Apr 2014 07:17 UTC**
Replying to [comment:2 adrpo]:
> > Technically it says that we need to merge the modifiers to A - not "apply it"
> 
> Well, the specification text is:
> 
> 5.6.1 Partial Flattening
> First the names of declared local classes and components are found. Here modifiers are merged to the local elements and redeclarations take effect (including redeclared elements, see Section 7.3). 
> Then base-classes are looked up, flattened and inserted into the class ...
> 
> 5.6.2 Flattening
> Partially flatten the class, apply the modifiers (Section 7.2) and flatten all local elements.
> 
> So 5.6.1 say that base-classes are looked up, flattened and 5.6.2 
> say that one has to partially flatten the class, apply the modifiers.
> 
> This is a bit wrong as partially flatten the class contains merging of modifiers.
> What does "apply the modifers" in 5.6.2 means then?

Well, merging is well-defined, so apply is something else and I agree that it needs to be better defined.

Since redeclarations are done earlier the remaining part is that apply means setting the values, i.e. for "x=5" when flattening "x" we apply the modifier to bind x to 5.

It would make more sense if redeclare would take effect in this step as well.

> Basically what you're saying is that modifiers are merged blindly without even checking that they exist (no lookup) until the flattening is done. 

Well, that would be one way of seeing it. Another is that some parts can be checked earlier - but some parts require that you flatten the model first.

Specifically:
There can be errors during merging - they could be reported then, or collected and reported later.
The modifier may modify something that cannot be modified - that can be detected either during flattening or afterwards.
But missing elements must be done after flattening, yes.

If you have 
```
 model M
   A a(x=a.y+b.y.x);
   B b(x=a.x+b.y.x, redeclare Y2 y);
 end M;
 M m(a.x.start=2);
```
You cannot know that A and B contain x and y until you have flattened them - and before that you cannot know that the references in the modifiers actually exists. 
You can still merge the new modifier.

**Comment by modelica-trac-importer** [comment]
**Comment by perost on 11 Apr 2014 13:03 UTC**
Ok, so consider this example instead:
```
model T1
  Real x = 1.0;
end T1;

model M
  model T2
    Real x = 2.0;
  end T2;
end M;

model A
  replaceable T1 t;
end A;

model B
  extends A(redeclare T2 t);
  extends M;
end B;
```
In this example, when should the redeclare modifier on A be applied? I assume that "T2 t" must be flattened in the scope of B, but that can't be done before flattening the extends since T2 is inherited from M. And to flatten the baseclass A it seems like we need to flatten the redeclaration first, since A should be flattened before inserting its contents into B. And indeed, Dymola says that T2 couldn't be found when I tried it. So is this example somehow illegal?

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 11 Apr 2014 15:33 UTC**
Replying to [comment:4 perost]:
> Ok, so consider this example instead:
> ```
> model T1
>   Real x = 1.0;
> end T1;
> 
> model M
>   model T2
>     Real x = 2.0;
>   end T2;
> end M;
> 
> model A
>   replaceable T1 t;
> end A;
> 
> model B
>   extends A(redeclare T2 t);
>   extends M;
> end B;
> ```
> In this example, when should the redeclare modifier on A be applied? 

The redeclare should take effect during partial flattening.

Thus we cannot even partially flatten the base-classes independently of each other and in any order and get the same result; so the example is illegal.
(Yes, there are some issues to improve in that chapter to clear up semantics without trying to change it. See #1397)

>I assume that "T2 t" must be flattened in the scope of B, but that can't be done before flattening the extends since T2 is inherited from M. And to flatten the baseclass A it seems like we need to flatten the redeclaration first, since A should be flattened before inserting its contents into B. And indeed, Dymola says that T2 couldn't be found when I tried it. So is this example somehow illegal?

Yes, currently.

However, it would be possible to specify that the redeclare is done later - I believe it would be cleaner and more general.

A current work-around is to use:
```
model B
  extends A;
  extends B;
  redeclare T2 t;
end B;
```
The disadvantage is that you cannot know where `t` is declared, and more importantly if there is an error and `t` is is missing you have no idea where it is missing from.

**Comment by modelica-trac-importer** [comment]
**Comment by jmattsson on 11 Apr 2014 15:49 UTC**
Replying to [comment:5 hansolsson]:
> A current work-around is to use:
> ```
> model B
>   extends A;
>   extends B;
>   redeclare T2 t;
> end B;
> ```

Not sure I follow correctly here. Did you mean this instead?
```mo
model B
  extends A;
  extends M;
  redeclare T2 t;
end B;
```

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 11 Apr 2014 15:51 UTC**
Replying to [comment:6 jmattsson]:
> Replying to [comment:5 hansolsson]:
> > A current work-around is to use:
> > ```
> > model B
> >   extends A;
> >   extends B;
> >   redeclare T2 t;
> > end B;
> > }}}
> Not sure I follow correctly here. Did you mean this instead?
> {{{#!mo
> model B
>   extends A;
>   extends M;
>   redeclare T2 t;
> end B;
> ```

Yes. Sorry - should have copy-pasted the working example instead of typing it again.

**Comment by modelica-trac-importer** [comment]
**Comment by perost on 11 Apr 2014 16:42 UTC**
Replying to [comment:5 hansolsson]:
> A current work-around is to use:
> ```
> model B
>   extends A;
>   extends M;
>   redeclare T2 t;
> end B;
> ```

Shouldn't that redeclare take effect at the same time as the redeclare in my example though? Or do you mean that redeclared elements should be applied after flattening the base classes?

**Comment by modelica-trac-importer** [comment]
**Comment by adrpo on 12 Apr 2014 08:14 UTC**
Replying to [comment:3 hansolsson]:
> Replying to [comment:2 adrpo]:
> 
> > Basically what you're saying is that modifiers are merged blindly without even checking that they exist (no lookup) until the flattening is done. 
> 
> Well, that would be one way of seeing it. Another is that some parts can be checked earlier - but some parts require that you flatten the model first.
> 
> Specifically:
> There can be errors during merging - they could be reported then, or collected and reported later.
> The modifier may modify something that cannot be modified - that can be detected either during flattening or afterwards.
> But missing elements must be done after flattening, yes.
> 

Or some parts are not checked at all (blindly merged/applied and discarded it seems).
You can try to flatten this model with Dymola and see how many errors you get:

```
model SelfReference1
  package P
    type T = Real(start = 1);
    package L
      replaceable package M = 
        P(T(start = 3, 
            somethingThatIsNotThere = 20, 
            SomethingElseThatIsNotThere(x = z)));

      model X
        parameter M.T t = 10;
      end X;
    end L;
  end P;

  P.L.X x;
end SelfReference1;
```

I would expect this model to give errors that:
- z is not present in P.L
- SomethingElseThatIsNotThere is not found in P.T
- somethingThatIsNotThere is not found in P.T

This is the danger of not checking earlier (where the modifiers are defined)
that their target or their value actually exists. I understand that this might
be complicated because of the way Modelica is defined but I still think it should
be possible to do.

**Comment by modelica-trac-importer** [comment]
**Comment by perost on 14 Apr 2014 10:22 UTC**
Here's another interesting model with regards to flattening:
```mo
model A
  replaceable model B
    model C
      Real x;
    end C;
  end B;

  B b;
end A;

model M
  extends A(redeclare model B = B.C);
end M;
```
It doesn't work in any tool I tested, so it's probably not a valid model. Still, I'm curious how this model is supposed to be flattened, even if it leads to an error. More specifically, at what point in the flattening is B.C looked up.

**Comment by modelica-trac-importer** [comment]
**Comment by eshmoylova on 14 Apr 2014 19:34 UTC**
Replying to [comment:5 hansolsson]:
> Replying to [comment:4 perost]:
> > Ok, so consider this example instead:
> > {{{
> > model T1
> >   Real x = 1.0;
> > end T1;
> > 
> > model M
> >   model T2
> >     Real x = 2.0;
> >   end T2;
> > end M;
> > 
> > model A
> >   replaceable T1 t;
> > end A;
> > 
> > model B
> >   extends A(redeclare T2 t);
> >   extends M;
> > end B;
> > }}}
> > In this example, when should the redeclare modifier on A be applied? 
> 
> The redeclare should take effect during partial flattening.
> 
> Thus we cannot even partially flatten the base-classes independently of each other and in any order and get the same result; so the example is illegal.

Unfortunately, there is no requirement in the Spec right now that says that we should be able to partially flatten the base-classes independently. There is a requirement (in 5.6.1) that the **lookup** of the base-classes should be independent of the order in which they are handled. The lookup seems to refer to looking up the name of the base class, and not the partial flattening. So there is nothing that says that this example is illegal.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 9 May 2014 08:12 UTC**
Replying to [comment:11 eshmoylova]:
> Unfortunately, there is no requirement in the Spec right now that says that we should be able to partially flatten the base-classes independently. There is a requirement (in 5.6.1) that the **lookup** of the base-classes should be independent of the order in which they are handled. The lookup seems to refer to looking up the name of the base class, and not the partial flattening. So there is nothing that says that this example is illegal.

I see it as a bit more complicated: The restriction is in the text but unfortunately part of the not very clear non-normative text in 5.6.1.

We can also make the example worse:
```
model M
  model M2
    Real x=2.0;
  end M2;
  replaceable T1 m;
end M;

model A
  model A3
    Real x=3.0;
  end A3;
  replaceable T1 a;
end A;

model B
  extends A(redeclare M2 a);
  extends M(redeclare A3 m);
end B;
```
The specification is not clear if we handle "extends A" or "extends M" first (the non-normative text states that it should not matter), but it is clear that we do one and then the other (since we first flatten and then insert the result - and before we insert the result lookup will not find anything) - unfortunately that is not possible in this example if we want the desired result.

There are two ways of handling this:
1. Add normative text corresponding to the comment about order not mattering; making this example illegal.
2. Split the description into more phases, such that the lookup of this modifier is done after adding the contents of the extends-clauses, something along the lines of:
5.6.1.1: First the names of declared local classes and components are found. Here modifiers are merged to the local elements. Then base-classes are looked up, (5.6.1.1 is applied to them) and inserted into the class. The lookup of the base-classes should be...
5.6.1.2. Redeclarations take effect (including redeclared elements, see Section 7.3). 
This also needs to be combined with the clarification of #1397.

**Comment by modelica-trac-importer** [comment]
**Comment by eshmoylova on 16 Jun 2014 03:08 UTC**
I would be in favor of the option 2. It will avoid situations when someone has a working model then while editing the code changes the order of the extends-clauses, and gets a broken model. If it gives an error it would be at least an easy thing to spot. But if, for example, there were top-level definitions of A3 and M2 in the example in the previous comment, there would be no error raised. Yet, if the top-level ones were different from the ones inside A and M changing the order of extends clauses would produce different results.

**Comment by modelica-trac-importer** [comment]
**Comment by jmattsson on 16 Jun 2014 11:36 UTC**
Replying to [comment:13 eshmoylova]:
> I would be in favor of the option 2. 

I agree. Also, that rule will be easier to understand, especially for users.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 17 Jun 2014 17:05 UTC**
Spec group:
Jesper: Breadth-first search of base-classes.
Hans:
For
```
model M
  model M2
    Real x=2.0;
  end M2;
  replaceable T1 m;
end M;

model A
  model A3
    Real x=3.0;
  end A3;
  replaceable T1 a;
end A;

model B
  extends A(redeclare M2 a);
  extends M(redeclare A3 m);
end B;
```
Step 1 looks up base-classes A and M
Step 2 finds that A contain model A3 and component a and merge modifier to a;
finds that M contain model M2 and component m and merge modifier to m
Step 3 combines contents of A and M into B

_We can now perform lookup in B_

Step 4 (in any order except for dependencies, and on demand) recurses for elements of B so that they are flattened which for components mean they get a type -  taking redeclare into account; (modifiers will use their original scope see #1397)

Martin (S): Duplicate elements is not specified here.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 17 Jun 2014 17:07 UTC**
Spec group:

Some are unsure about adding this to 3.3r1 and/or the implications.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 18 Jun 2014 09:32 UTC**
Turning the above into spec text:

5.6.1 Partial Flattening
First the names of declared local classes and components are found. Here modifiers are merged to the local elements. Then base-classes are looked up, partially flattened (in the same way) and inserted into the class.

After the partial flattening it is possible to perform lookup in the class, since the name of all local and inherited elements are known. 
5.6.2. Flattening
Partially flatten all elements of the class. Here classes for components are looked up (taking redeclarations into account).

**Comment by modelica-trac-importer** [comment]
**Comment by perost on 18 Jun 2014 10:02 UTC**
Replying to [comment:17 hansolsson]:
> Turning the above into spec text:
> ... 

Maybe I'm missing something, but doesn't this say that inherited elements should be flattened in the local scope? I.e. base-classes are partially flattened and inserted into the class, and then both local and inherited elements are partially flattened in the flattening step. Is that really the intent?

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 18 Jun 2014 10:47 UTC**
Revised 5.6.2 based on discussion at meeting:
All elements are flattened, starting by partially flattening them.

As part of this classes for component elements are looked up, i.e. modifiers including redeclarations for component take effect – and the flattened components are inserted into the resulting class.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 18 Jun 2014 10:49 UTC**
Replying to [comment:18 perost]:
> Replying to [comment:17 hansolsson]:
> > Turning the above into spec text:
> > ... 
> Maybe I'm missing something, but doesn't this say that inherited elements should be flattened in the local scope? I.e. base-classes are partially flattened and inserted into the class, and then both local and inherited elements are partially flattened in the flattening step. Is that really the intent?

This is related to #1397 - which also should be included. The idea is that they are inserted, but keep their scope (which needs to be clarified further).

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 24 Jun 2014 11:16 UTC**
These tickets are important, but cannot be resolved for 3.3r1.

**Comment by modelica-trac-importer** [comment]
**Comment by perost on 4 Sep 2014 11:36 UTC**
Replying to [comment:17 hansolsson]:
Has there been any progress on this issue? I realize that it probably won't be resolved until at least the next design meeting, but it would nonetheless be good to have a proposal that could be discussed before the meeting. This is a complex issue that I find very important to resolve asap, and if there's no clear idea on how to fix it before the next meeting I fear it will be delayed even further.

Anyway, a question about this:
> Turning the above into spec text:
> 
> 5.6.1 Partial Flattening
> First the names of declared local classes and components are found. Here modifiers are merged to the local elements. Then base-classes are looked up, partially flattened (in the same way) and inserted into the class.
> 
> After the partial flattening it is possible to perform lookup in the class, since the name of all local and inherited elements are known. 

Since lookup in the class is not possible until partial flattening is done, does that mean that classes inside local classes can't be extended from? I.e.:
```
model A
  extends B.C; // B.C can't be found until A has been partially flattened?

  model B
    model C
    end C;
  end B;
end A; 
```

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 4 Sep 2014 16:35 UTC**
Replying to [comment:22 perost]:
> Replying to [comment:17 hansolsson]:
> Has there been any progress on this issue? I realize that it probably won't be resolved until at least the next design meeting, but it would nonetheless be good to have a proposal that could be discussed before the meeting. This is a complex issue that I find very important to resolve asap, and if there's no clear idea on how to fix it before the next meeting I fear it will be delayed even further.
> 
> Anyway, a question about this:
> > Turning the above into spec text:
> > 
> > 5.6.1 Partial Flattening
> > First the names of declared local classes and components are found. Here modifiers are merged to the local elements. Then base-classes are looked up, partially flattened (in the same way) and inserted into the class.
> > 
> > After the partial flattening it is possible to perform lookup in the class, since the name of all local and inherited elements are known. 
> Since lookup in the class is not possible until partial flattening is done, does that mean that classes inside local classes can't be extended from? I.e.:

It means that the procedure have to be described in more detail, it is described as being a two-part procedure ("First" and "Then") but they need to be separated more.

So the dead-lock above is avoided since the lookup of base-classes in 5.6.1 ignores inherited classes (good since inheritance is not yet done). Afterwards base-classes are looked up again - and it is an error if this lookup gives a different result (etc as previously discussed).

Basically 5.6.1 should be split into these two steps: 
* First finding local names (and merging modifiers).
* Then adding base-class contents.
And the first step should have a name.

Slightly updated example:
```
 model A
   extends B.C; // B.C can't be found until A has been partially flattened?
 
   model B
     model C
       extends D; // To make it fun!
     end C;
   end B;
 end A; 
 model D "Nothing to see here"
   ...
 end D;
```
When partially flattening A:
1. We find that A contains a model B and an extends class.
2. Normal lookup of B.C is comprised of finding B in A or its enclosing scopes, and then a C inside this B.
3. Lookup of B in A finds the local class B.
4. We now have to flatten B (will have to verify exactly how much is needed) to find if it contains a C.

after a bit we then begin to partially flatten C:
6. We find that C contains an extends class.
7. We try to find D in A.B.C and go out the enclosing scopes until we find D (at the top-level), and inserts its contents to C.

8. Now we can inserts the contents of (B.)C to A.
9. And finally the partial flattening of A can be completed, and lookup of B.C will still give the same result as before (unless there is a local "B" in "D" - which would be an error; similarly if there is a D in D it would also be an error - but in C).

Thus the specification text would be something like:
5.6.1.1: First the names of declared local classes and components are found. Here modifiers are merged to the local elements. 
5.6.1.2. Then base-classes are looked up, (5.6.1 is applied to them) and inserted into the class. The lookup of the base-classes should be…
5.6.2. All elements are flattened, starting by partially flattening them.
As part of this classes for component elements are looked up, i.e. modifiers including redeclarations for component take effect and the flattened components are inserted into the resulting class.

After step 5.6.1 it is possible to perform lookup in the class, since all named elements are known. It is sometimes necessary to performing lookup before step 5.6.1.2 is complete to handle base-classes, but the result needs to be verified after 5.6.1.2 is complete.

(Step 5.6.2 should also be reformulated, so that there is one step that modifiers take effect and a final step that will ensure that all elements are flattened (some may already have been flattened).)

Most OO-languages avoid this by having the base-classes outside the class itself, which normally prevents inheritance from local class and require a special syntax for separating base-classes from the rest of the contents, and changes to the base-classes are done by overriding virtual functions instead of using modifiers on the extends-clauses. (Import-statements also complicate this, if we want the import-statement to be class-specific, and still be used for finding base-classes.)

**Comment by modelica-trac-importer** [comment]
**Modified by perost on 10 Dec 2014 14:46 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by kurzbach on 23 Feb 2015 16:03 UTC**
The attachment contains a try to specify the instantiation process in a more algorithmic way, and to split it in multiple stages. Finally it is completed with the generation of the flat equation system.

Hans, Please have a look at an comment it.

**Comment by modelica-trac-importer** [comment]
**Comment by perost on 9 Mar 2015 14:15 UTC**
Replying to [comment:25 kurzbach]:
> The attachment contains a try to specify the instantiation process in a more algorithmic way, and to split it in multiple stages. Finally it is completed with the generation of the flat equation system.

I have a couple of comments on this proposal:

**5.3 Static Name Lookup:**
You write <from specification>, but I assume that the mentions of flattening regarding composite name lookup should be changed to say instantiation instead.

**5.6.1 Instantiation Process:**
One of the inputs to the instantiation is a class' environment. I think how this environment is constructed also needs to be specified. For example: the top level classes are added to the top level scope, and the class to be instantiated is then looked up in this scope following the normal lookup rules.

**Partial Instantiation, Instantiation:**
The text alternates between using instantiation and flattening. I'm assuming this is just a typo, and it should be instantiation everywhere in these two sections.
  
**Partial Instantiation:**
>First local classes are found, redeclarations referring to them are applied, modifications are merged (see 7.2.3), and they are inserted into the current instance and then partially instantiated.

I think it might be better to slightly reorder the steps so that extends are partially instantiated before local classes, to handle local classes which depend on the local extends-clauses. It's probably possible to handle with the proposed order, but I find this order more natural:
  1. Local classes are found, etc.
  2. Extends are partially flattened.
  3. Local classes are partially flattened.

>When currently an extends clause is partially flattened, then classes which are redeclared by a class extends redeclaration (see 7.3.1) are ignored.

This is ambiguous. Should those classes be ignored while looking up the type of an extends clause, or while partially flattening that type? Also, I'm not sure what "When currently" refers to. It seems like this should come after the next part about partially flattening extends, or have I misunderstood what this sentence refers to?

>Then types of extends clauses are looked up, the local classes inside of them are partially flattened in the context of the current instance. 

"partially flattened in the context of the current instance" seems ambiguous to me. I would rather say that the types of extends clauses are looked up and partially flattened, and the content (or just classes?) of those partially flattened classes are inserted into the current instance. It should probably be mentioned somewhere that each class knows its scope, similar to modifiers. I.e. inherited classes should be instantiated in the scope they came from, and not the current scope.

**Instantiation**
>Then the classes of the components are instantiated in the context of the created instance using the merged modifications.

"the classes of the components" could either mean the classes inside the components, or the types of the components. Maybe rewrite as: "Then each component's type is instantiated ...".

**Comment by modelica-trac-importer** [comment]
**Comment by kurzbach on 10 Mar 2015 06:42 UTC**
Replying to [comment:26 perost]:
> Replying to [comment:25 kurzbach]:
> > The attachment contains a try to specify the instantiation process in a more algorithmic way, and to split it in multiple stages. Finally it is completed with the generation of the flat equation system.
> I have a couple of comments on this proposal:
> 
> **5.3 Static Name Lookup:**
> You write <from specification>, but I assume that the mentions of flattening regarding composite name lookup should be changed to say instantiation instead.
> 

That's true.

> **5.6.1 Instantiation Process:**
> One of the inputs to the instantiation is a class' environment. I think how this environment is constructed also needs to be specified. For example: the top level classes are added to the top level scope, and the class to be instantiated is then looked up in this scope following the normal lookup rules.

Yes, that needs to be added.

> 
> **Partial Instantiation, Instantiation:**
> The text alternates between using instantiation and flattening. I'm assuming this is just a typo, and it should be instantiation everywhere in these two sections.
>   

Yes, you are right. There are many places in the document where "partial flattening" is mentioned. This is wrong and should be "partial instantiation". I see "flattening" as the global process, where as instantiation and partial instantiation are steps within.

> **Partial Instantiation:**
> >First local classes are found, redeclarations referring to them are applied, modifications are merged (see 7.2.3), and they are inserted into the current instance and then partially instantiated.
> I think it might be better to slightly reorder the steps so that extends are partially instantiated before local classes, to handle local classes which depend on the local extends-clauses. It's probably possible to handle with the proposed order, but I find this order more natural:
>   1. Local classes are found, etc.
>   2. Extends are partially flattened.
>   3. Local classes are partially flattened.
> 

Which order you prefer? extends clauses before or after local classes? In the spec there is the rule, that extends clauses can be independendly instantiated from the elements of the surrounding class. So probably having extends clauses first would be the right way. So maybe I have to change the document, but it is also not really clear to me.

> >When currently an extends clause is partially flattened, then classes which are redeclared by a class extends redeclaration (see 7.3.1) are ignored.
> This is ambiguous. Should those classes be ignored while looking up the type of an extends clause, or while partially flattening that type? Also, I'm not sure what "When currently" refers to. It seems like this should come after the next part about partially flattening extends, or have I misunderstood what this sentence refers to?

This means, when during the partial instantiation of an extends clause a local class is to be partially instantiated and there is a class extends redeclaration in the environment, the this class is ignored, i.e. not instantiated, because it is instantiated in the scope, where the class extends belongs to. Else there would be multiple instances of the same class.

> 
> >Then types of extends clauses are looked up, the local classes inside of them are partially flattened in the context of the current instance. 
> "partially flattened in the context of the current instance" seems ambiguous to me. I would rather say that the types of extends clauses are looked up and partially flattened, and the content (or just classes?) of those partially flattened classes are inserted into the current instance.

Ok. 

> It should probably be mentioned somewhere that each class knows its scope, similar to modifiers. I.e. inherited classes should be instantiated in the scope they came from, and not the current scope.

That's right.
> 
> **Instantiation**
> >Then the classes of the components are instantiated in the context of the created instance using the merged modifications.
> "the classes of the components" could either mean the classes inside the components, or the types of the components. Maybe rewrite as: "Then each component's type is instantiated ...".

Ok.

I will make a second version of the document and take these comments into account, but I still wait for comments from Hans before doing that.

**Comment by modelica-trac-importer** [comment]
**Comment by perost on 10 Mar 2015 09:58 UTC**
Replying to [comment:27 kurzbach]:
> Which order you prefer? extends clauses before or after local classes? In the spec there is the rule, that extends clauses can be independendly instantiated from the elements of the surrounding class. So probably having extends clauses first would be the right way. So maybe I have to change the document, but it is also not really clear to me.

I was thinking of cases such as this:
```mo
model M
  model A
    model B
      ...
    end B;
  end A;

  extends A;  // Inherits B

  model C
    B b;  // Using inherited B
  end C;
end M;
```
To partially instantiate C in this case we need to know what's been inherited from A. So then the lookup needs to be able to look things up in extends-clauses as well as local elements, if we partially instantiate extends clauses first. On the other hand, if we add the local classes to the instance first and partially instantiate extends before the local classes, then the lookup becomes less complicated since it doesn't need to handle extends clauses. 

The specification currently says in 5.6.1: "Local classes used for extends should be possible to flatten before inserting the result of flattening the extends-clauses". Is this the rule you're referring to? So it should be possible to look up a local class and then partially instantiate it when partially instantiating extends clauses (i.e. M can't extend C). But yeah, it's not really clear to me either if this works in all cases.

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 10 Mar 2015 10:05 UTC**
My understanding of the locally defined superclasses is that you will always be able to tell them apart (since extends-clauses only use names you can select all required (for the superclass instantiation) local classes and handle them before the extends-clause. The spec requires that these classes can be instantiated before the outer extends-clause.

However, the spec does not mention the granularity here. Consider this case:

```mo
class A
  extends B.C;
  class B
    class C type T = Real; end C;
    class D type S = T; /* = A.T inherited in A above */ end D;
  end B;
end A;
```

Here, it is clearly possible to handle the classes in the following order: A.B.C, (extends of) A, A.B.D, A.B, A. However, the spec does not really say that this is standard behavior.

**Comment by modelica-trac-importer** [comment]
**Comment by perost on 11 Mar 2015 11:57 UTC**
After thinking a bit more I think I need to revise my suggestion a bit. Modifying Hans' model in comment:12 to redeclare classes instead of components gives this:
```mo
model M
  model A
    model A2
      Real x = 1.0;
    end A2;

    replaceable model A3 = A2;
    A3 a3;
  end A;

  model B
    model B2
      Real x = 2.0;
    end B2;
    replaceable model B3 = B2;
    B3 b3;
  end B;

  extends A(redeclare model A3 = B2);
  extends B(redeclare model B3 = A2);
end M;
```
According to my previous suggestion we should first add the local classes A and B, and then partially instantiate the extends. If I've understood things correctly this involves applying the redeclares. In that case it seems that a slightly different order of doing things might be better:
  1. Local classes are inserted into the current instance.
  2. Base classes are looked up, and their classes are inserted into the current instance.
  3. Redeclarations are applied.
  4. Both local classes and classes from extends are partially flattened.

Also, Dymola can't handle this model, it says "Missing base class B2" on extends A. So another solution would be to just say that this kind of model is wrong. It could be argued that the rule about local classes being used for extends applies also in this case, since "model A3 = B2" is a kind of extends even though it's in a modifier.

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 11 Mar 2015 17:49 UTC**
Replying to [comment:30 perost]:

> Also, Dymola can't handle this model, it says "Missing base class B2" on extends A. So another solution would be to just say that this kind of model is wrong. It could be argued that the rule about local classes being used for extends applies also in this case, since "model A3 = B2" is a kind of extends even though it's in a modifier.

In your example, to resolve B2 in the context M (which is basically what this short-hand-style of redeclaration does according to my understanding), you would have to resolve the extends clause. However, the spec says explicitly that the order of extends-clauses must not matter. So the model **is** wrong, regardless of the implementation.

Also please note that your algorithm lacks in point 2: When you insert only local classes without flattening them, you might not be able to find all base classes:

```mo
class A
  extends B.C;
end A;

class B
  extends D;
end B;

class D class C end C; end D;
```

AFAIK any tool needs to do a dependency analysis to figure out in which order classes shall be flattened to decide where to look for base classes. In my example above, this dependency is quite simple, however it might be a lot more complicated (e.g. there might be multiple possible sources). I am currently working on an implementation of said dependency analysis and intend to submit the formalisation to the Modelica Conference. Maybe you are interested in co-authoring?

**Comment by modelica-trac-importer** [comment]
**Comment by perost on 11 Mar 2015 18:23 UTC**
Replying to [comment:31 choeger]:
> Replying to [comment:30 perost]:
> 
> > Also, Dymola can't handle this model, it says "Missing base class B2" on extends A. So another solution would be to just say that this kind of model is wrong. It could be argued that the rule about local classes being used for extends applies also in this case, since "model A3 = B2" is a kind of extends even though it's in a modifier.
> 
> In your example, to resolve B2 in the context M (which is basically what this short-hand-style of redeclaration does according to my understanding), you would have to resolve the extends clause. However, the spec says explicitly that the order of extends-clauses must not matter. So the model **is** wrong, regardless of the implementation.

The order of the extends clauses wouldn't matter in this case. When partially instantiating M we first find A and B and add them to the instance of M. Then we look up e.g. extends A, find that it contains A2 and A3. We then merge the redeclare of A3 (this wasn't clear in my algorithm), and add A2 and A3 to M. We don't need to know what B2 is in this step. Then we do the same with extends B. Then finally we partially flatten all of the classes. In this step we apply the redeclares of A3 and B3, but since B2 and A2 have been added to the scope already we can look them up, and the order we handle the extends in doesn't matter.

> Also please note that your algorithm lacks in point 2: When you insert only local classes without flattening them, you might not be able to find all base classes:
> 
> ```
> #!mo
> class A
>   extends B.C;
> end A;
> 
> class B
>   extends D;
> end B;
> 
> class D class C end C; end D;
> ```

I don't see any issues with this model. To partially instantiate A:
  * No local classes in A.
  * Look up base class B.C:
    * Look up B in current scope -> found in top scope.
    * Instantiate B:
      * Partially instantiate B:
        * No local classes in B.
        * Look up extends D -> found in top scope.
        * Add classes from D to B -> C is added to B.
        * Partially instantiate classes in B.
      * Fully instantiate B -> nothing to do since no components.
    * Look up C in B -> found in B.
  * Add classes from B.C in A -> nothing is added to A.
  * Partially instantiate classes in A.

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 11 Mar 2015 20:06 UTC**
Replying to [comment:32 perost]:
> The order of the extends clauses wouldn't matter in this case. When partially instantiating M we first find A and B and add them to the instance of M.

That is a matter of definition: To me the extends clauses contain the modification (i.e. you are inheriting the class after application of redeclaration in the local scope). Your view seems to be equivalent to the "long-style" redeclaration (i.e. having the redeclaration as an element of the class). If you view both as equivalent, however, your algorithm should probably rewrite the short form into the long form to avoid confusion (I believe however that they are not equivalent and B2 has to be found locally).

> I don't see any issues with this model. To partially instantiate A:
>   \* No local classes in A.
>   \* Look up base class B.C:
>     \* Look up B in current scope -> found in top scope.
>     \* Instantiate B:

That is precisely the point: Your algorithm in comment 30 did not contain this kind of recursion. To resolve the base class, your instantiation algorithm **needs** to be recursive, hence you might either end up with endless recursion or you need some tool to prevent this. I implemented a dependency analysis (and found that recursive models are already in the MSL, but that's another ticket).

**Comment by modelica-trac-importer** [comment]
**Comment by perost on 11 Mar 2015 21:39 UTC**
Replying to [comment:33 choeger]:
> Replying to [comment:32 perost]:
> > The order of the extends clauses wouldn't matter in this case. When partially instantiating M we first find A and B and add them to the instance of M.
> 
> That is a matter of definition: To me the extends clauses contain the modification (i.e. you are inheriting the class after application of redeclaration in the local scope). Your view seems to be equivalent to the "long-style" redeclaration (i.e. having the redeclaration as an element of the class). If you view both as equivalent, however, your algorithm should probably rewrite the short form into the long form to avoid confusion (I believe however that they are not equivalent and B2 has to be found locally).

No, of course they are not equivalent. Redeclares as modifiers are only applied to the extends clause they modify, while redeclares as elements are applied to all inherited elements. I.e. merging redeclares as modifiers is done in step 2 while adding the classes from the extends, and redeclares as elements in step 3 when all inherited classes have been added. The merged redeclares are then applied when the local and inherited classes are partially flattened.
> 
> > I don't see any issues with this model. To partially instantiate A:
> >   \* No local classes in A.
> >   \* Look up base class B.C:
> >     \* Look up B in current scope -> found in top scope.
> >     \* Instantiate B:
> 
> That is precisely the point: Your algorithm in comment 30 did not contain this kind of recursion. To resolve the base class, your instantiation algorithm **needs** to be recursive, hence you might either end up with endless recursion or you need some tool to prevent this. I implemented a dependency analysis (and found that recursive models are already in the MSL, but that's another ticket).
> 

Do you mean the instantiation of B? That's how the lookup of composite names is currently specified to work, so it doesn't need to be part of the algorithm.

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 11 Mar 2015 22:10 UTC**
Replying to [comment:34 perost]:

> No, of course they are not equivalent. Redeclares as modifiers are only applied to the extends clause they modify, while redeclares as elements are applied to all inherited elements. 

But there can exist always only one superclass for each inherited element, so _all inherited elements_ in that case is at most one element, right?

> Do you mean the instantiation of B? That's how the lookup of composite names is currently specified to work, so it doesn't need to be part of the algorithm.

In that case, your algorithmic specification should also contain the lookup. It is bad style IMO to simply hide such an important feat as recursion (and thus a source of non-termination).

**Comment by modelica-trac-importer** [comment]
**Comment by perost on 11 Mar 2015 23:20 UTC**
Replying to [comment:35 choeger]:
> Replying to [comment:34 perost]:
> 
> > No, of course they are not equivalent. Redeclares as modifiers are only applied to the extends clause they modify, while redeclares as elements are applied to all inherited elements. 
> 
> But there can exist always only one superclass for each inherited element, so _all inherited elements_ in that case is at most one element, right?

Not really:
```mo
class A
  replaceable class M end M;
end A;

class B
  replaceable class M end M;
end B;

class C
  extends A;
  extends B;
  redeclare class M end M;
end C;
```
My understanding is that the redeclare of M in C applies to both of the inherited elements, although this is not really clear in the specification.

> > Do you mean the instantiation of B? That's how the lookup of composite names is currently specified to work, so it doesn't need to be part of the algorithm.
> 
> In that case, your algorithmic specification should also contain the lookup. It is bad style IMO to simply hide such an important feat as recursion (and thus a source of non-termination).

The instantiation and lookup depend on each other, so they can be considered as one large algorithm. The lookup is already defined, so I see no need to take it into consideration when defining the instantiation. And the instantiation is already recursive anyway.

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 12 Mar 2015 09:09 UTC**
Replying to [comment:36 perost]:
> My understanding is that the redeclare of M in C applies to both of the inherited elements, although this is not really clear in the specification.

Yeah, but in that case the spec also demands that both are the **same**, so that particular corner case does not really matter IMO. The important point is when and where to evaluate the rhs of a redeclaration. I opened #1680 for that issue.
 
> The instantiation and lookup depend on each other, so they can be considered as one large algorithm. The lookup is already defined, so I see no need to take it into consideration when defining the instantiation. And the instantiation is already recursive anyway.

As I said, I consider it slightly bad style. The instantiation you discussed (recursively) traversed a finite structure and hence would always terminate. Considering the lookup-recursion, the termination cannot be guaranteed anymore (and in fact the instantiation will not terminate for several cases). My understanding is that most users and even some developers where surprised to learn about non-terminating behavior during flattening. So I think the issue should be made explicit.

**Comment by modelica-trac-importer** [comment]
**Comment by perost on 12 Mar 2015 09:26 UTC**
Replying to [comment:37 choeger]:
> Replying to [comment:36 perost]:
> > My understanding is that the redeclare of M in C applies to both of the inherited elements, although this is not really clear in the specification.
> 
> Yeah, but in that case the spec also demands that both are the **same**, so that particular corner case does not really matter IMO. The important point is when and where to evaluate the rhs of a redeclaration. I opened #1680 for that issue.

On the other hand it would seem that applying element redeclares before checking for equivalent elements would be more useful. That would mean that it would be possible to inherit several elements with the same name but possibly different declarations, and replace all of them to be compatible with one redeclare instead of duplicating the redeclare as a modifier for each extends. But maybe that's not desirable, I don't know.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 15 Apr 2015 14:33 UTC**
Ticket retargeted after milestone closed
